### PR TITLE
Unit tests for MixinScheduler

### PR DESCRIPTION
**Pull request by hysz**
## Description

This PR includes unit tests for `MixinScheduler` in the Staking Contracts.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/staking/contracts/test/TestMixinScheduler.sol [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        //     `epochEndTime = currentEpochStartTimeInSeconds + epochDurationInSeconds`
+        // (ii) We want adjust internal state such that:
+        //      `epochEndTime - block.timestamp = epochEndTimeDelta`, or
+        //      `currentEpochStartTimeInSeconds + epochDurationInSeconds - block.timestamp = epochEndTimeDelta`
+        //
+        // To do this, we:
+        //  (i) Set `epochDurationInSeconds` to a constant value of 1, and
+        //  (ii) Rearrange the eqn above to get:
+        //      `currentEpochStartTimeInSeconds = epochEndTimeDelta + block.timestamp - epochDurationInSeconds`
+        epochDurationInSeconds = 1;
```
Comment (mzhu25) [inline contracts/staking/contracts/test/TestMixinScheduler.sol]: why do we set this to 1? can we use a more realistic value?

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+            // Mine the block that this tx would've been in.
+            await env.web3Wrapper.mineBlockAsync();
+            const blockNumber = await env.web3Wrapper.getBlockNumberAsync();
+            const blockTimestampAsNumber = await env.web3Wrapper.getBlockTimestampAsync(blockNumber);
+            const blockTimestamp = new BigNumber(blockTimestampAsNumber);
+            const epochEndTime = blockTimestamp.plus(epochEndTimeDelta);
+            return expect(tx).to.revertWith(
+                new StakingRevertErrors.BlockTimestampTooLowError(epochEndTime, blockTimestamp),
+            );
```
Comment (mzhu25) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: nice

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            const epochEndTimeDelta = new BigNumber(-10);
+            const txReceipt = await testContract.goToNextEpochTest.awaitTransactionSuccessAsync(epochEndTimeDelta);
+            // tslint:disable-next-line no-unnecessary-type-assertion
+            const testLog: TestMixinSchedulerGoToNextEpochTestInfoEventArgs = (txReceipt.logs[0] as LogWithDecodedArgs<
+                TestMixinSchedulerGoToNextEpochTestInfoEventArgs
+            >).args;
+            const currentEpoch = await testContract.currentEpoch.callAsync();
+            const currentEpochStartTimeInSeconds = await testContract.currentEpochStartTimeInSeconds.callAsync();
+            expect(currentEpoch).to.bignumber.equal(testLog.oldEpoch.plus(1));
+            expect(currentEpochStartTimeInSeconds).to.bignumber.equal(testLog.blockTimestamp);
```
Comment (mzhu25) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: optional nit: use `verifyEventsFromLogs` for these events

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        it('Should revert if epoch end time is strictly greater than block timestamp', async () => {
+            const epochEndTimeDelta = new BigNumber(10);
+            const tx = testContract.goToNextEpochTest.awaitTransactionSuccessAsync(epochEndTimeDelta);
+            try {
+                await tx;
+                // tslint:disable-next-line no-empty
+            } catch (e) {}
+
+            // Mine the block that this tx would've been in.
+            await env.web3Wrapper.mineBlockAsync();
```
Comment (jalextowle) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: This is a pretty cool pattern. Might make sense to make a function and put in test-utils, but optional for this PR.

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    let accounts: string[];
+    let owner: string;
+    let authorizedAddress: string;
+    let notAuthorizedAddresses: string[];
+    let testContract: TestMixinSchedulerContract;
+
+    before(async () => {
+        // Create accounts
+        accounts = await env.getAccountAddressesAsync();
+        [owner, authorizedAddress, ...notAuthorizedAddresses] = accounts;
```
Comment (jalextowle) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: It turns out that none of these accounts are actually being used. Let's remove them and remove all of the `let` variables above.

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
+import { blockchainTests, expect } from '@0x/contracts-test-utils';
+import { StakingRevertErrors } from '@0x/order-utils';
+import { BigNumber } from '@0x/utils';
+import { LogWithDecodedArgs } from 'ethereum-types';
+import * as _ from 'lodash';
```
Comment (jalextowle) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: This seems like it isn't used. Can we remove for now?

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            const currentEpochEarliestEndTimeInSeconds = await testContract.getCurrentEpochEarliestEndTimeInSeconds.callAsync();
+            expect(currentEpochEarliestEndTimeInSeconds).to.bignumber.equal(
+                expectedCurrentEpochEarliestEndTimeInSeconds,
+            );
+        });
+    });
+
+    describe('_initMixinScheduler', () => {
+        it('Should succeed if scheduler is not yet initialized (`currentEpochStartTimeInSeconds == 0`)', async () => {
+            const initCurrentEpochStartTimeInSeconds = new BigNumber(0);
```
Comment (jalextowle) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: Optional Nit: `new BigNumber(0)` ==> `constants.ZERO_AMOUNT

File: contracts/staking/contracts/test/TestMixinScheduler.sol (hunk 1 above)
Comment (hysz) [reply contracts/staking/contracts/test/TestMixinScheduler.sol]: The value just needs to be non-zero. Allows us to verify that the `endTime` is accounting for both `currentEpochStartTimeInSeconds ` and `epochDurationInSeconds`.

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            const testLog: TestMixinSchedulerGoToNextEpochTestInfoEventArgs = (txReceipt.logs[0] as LogWithDecodedArgs<
+                TestMixinSchedulerGoToNextEpochTestInfoEventArgs
+            >).args;
+            const currentEpoch = await testContract.currentEpoch.callAsync();
+            const currentEpochStartTimeInSeconds = await testContract.currentEpochStartTimeInSeconds.callAsync();
+            expect(currentEpoch).to.bignumber.equal(testLog.oldEpoch.plus(1));
+            expect(currentEpochStartTimeInSeconds).to.bignumber.equal(testLog.blockTimestamp);
+        });
+
+        it('Should succeed if epoch end time is equal to block timestamp', async () => {
```
Comment (jalextowle) [inline contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: Rigor lvl 💯

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts (hunk 6 above)
Comment (hysz) [reply contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: Good eye!

File: contracts/staking/test/unit_tests/mixin_scheduler_test.ts (hunk 4 above)
Comment (hysz) [reply contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: 🙏 - although it's such an edge case idk if it would be worth abstracting as-is.

Comment (jalextowle) [reply contracts/staking/test/unit_tests/mixin_scheduler_test.ts]: Fair enough lol
